### PR TITLE
Implement initial CI configuration

### PR DESCRIPTION
**Pull request by nathan-weinberg**
Supersedes #7 

This PR adds the following CI jobs:
- Dependabot
- GitHub Action linting
- Markdown Linting (with fixes)
- Spellcheck (with fixes)
- Code Linting (with **some** fixes - see note below)

As well as a `Makefile` for local running of the checks.

**Additional Notes**
- I will do a followup for PyPI packaging but I want to work directly with the folks from Training Squad on that to keep them in the loop
- WRT the Python checks - @RobotSail @Maxusmusti @JamesKunstle I didn't go through all the failures (the fixes I did were auto-format fixes that have no functional impact) because I don't want to hamper y'alls velocity and break things currently working - so the checks remain but are non-blocking for now - I would encourage y'all to address these when it makes sense

### CODE REVIEW

**Comment by russellb** [comment]
know why the jobs aren't running on this PR? I'd prefer to see them run green before merging.

**Comment by bjhargrave** [comment]
> know why the jobs aren't running on this PR?

I don't think github will run new actions in a PR since that is a security risk.

**Comment by cdoern** [comment]
> I will do a followup for PyPI packaging but I want to work directly with the folks from Training Squad on that to keep them in the loop

any chance we can do this today @nathan-weinberg ? I'd really like to test this out

**Comment by nathan-weinberg** [comment]
> know why the jobs aren't running on this PR? I'd prefer to see them run green before merging.

They won't run until they are merged - we saw this in https://github.com/instructlab/eval/pull/1 - but these are a carbon-copy from the current `eval` config so I'm confident in them

File: .github/workflows/matchers/actionlint.dockerfile [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Since dependabot cannot update workflows using docker,
```
Comment (bjhargrave) [inline .github/workflows/matchers/actionlint.dockerfile:1]: This is in the wrong folder. It needs to be in .github/workflows.

**Comment by nathan-weinberg** [comment]
> > I will do a followup for PyPI packaging but I want to work directly with the folks from Training Squad on that to keep them in the loop
> 
> any chance we can do this today @nathan-weinberg ? I'd really like to test this out

Let's chat offline

**Comment by russellb** [comment]
for the Python linting issues you call out -- can you just update the linter config so that it passes?  The typical approach I'd prefer is always have a config that runs green, but work over time to improve the config so more rules can be enforced.

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+env:
+  PYTHON_VERSION: 3.11
+
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
```
Comment (bjhargrave) [inline .github/workflows/lint.yml:37]: This is not hardened. See the newer lint workflow from the cli repo.

**Comment by cdoern** [comment]
```
ERROR: Ignored the following versions that require a different python version: 0.52.0 Requires-Python >=3.6,<3.9; 0.52.0rc3 Requires-Python >=3.6,<3.9; 0.53.0 Requires-Python >=3.6,<3.10; 0.53.0rc1.post1 Requires-Python >=3.6,<3.10; 0.53.0rc2 Requires-Python >=3.6,<3.10; 0.53.0rc3 Requires-Python >=3.6,<3.10; 0.53.1 Requires-Python >=3.6,<3.10; 0.54.0 Requires-Python >=3.7,<3.10; 0.54.0rc2 Requires-Python >=3.7,<3.10; 0.54.0rc3 Requires-Python >=3.7,<3.10; 0.54.1 Requires-Python >=3.7,<3.10; 0.55.0 Requires-Python >=3.7,<3.11; 0.55.0rc1 Requires-Python >=3.7,<3.11; 0.55.1 Requires-Python >=3.7,<3.11; 0.55.2 Requires-Python >=3.7,<3.11
ERROR: Could not find a version that satisfies the requirement dolomite-engine (from instructlab-training) (from versions: none)
ERROR: No matching distribution found for dolomite-engine
```
Found this when pip installing, I think this will fail the linter as well

**Review by bjhargrave — CHANGES_REQUESTED** [review]

**Comment by nathan-weinberg** [comment]
> ```
> ERROR: Ignored the following versions that require a different python version: 0.52.0 Requires-Python >=3.6,<3.9; 0.52.0rc3 Requires-Python >=3.6,<3.9; 0.53.0 Requires-Python >=3.6,<3.10; 0.53.0rc1.post1 Requires-Python >=3.6,<3.10; 0.53.0rc2 Requires-Python >=3.6,<3.10; 0.53.0rc3 Requires-Python >=3.6,<3.10; 0.53.1 Requires-Python >=3.6,<3.10; 0.54.0 Requires-Python >=3.7,<3.10; 0.54.0rc2 Requires-Python >=3.7,<3.10; 0.54.0rc3 Requires-Python >=3.7,<3.10; 0.54.1 Requires-Python >=3.7,<3.10; 0.55.0 Requires-Python >=3.7,<3.11; 0.55.0rc1 Requires-Python >=3.7,<3.11; 0.55.1 Requires-Python >=3.7,<3.11; 0.55.2 Requires-Python >=3.7,<3.11
> ERROR: Could not find a version that satisfies the requirement dolomite-engine (from instructlab-training) (from versions: none)
> ERROR: No matching distribution found for dolomite-engine
> ```
> 
> Found this when pip installing, I think this will fail the linter as well

@RobotSail @Maxusmusti @JamesKunstle any idea why this is happening? I'm hitting the same issue

**Review by bjhargrave — APPROVED** [review]

**Comment by russellb** [comment]
> > ```
> > ERROR: Ignored the following versions that require a different python version: 0.52.0 Requires-Python >=3.6,<3.9; 0.52.0rc3 Requires-Python >=3.6,<3.9; 0.53.0 Requires-Python >=3.6,<3.10; 0.53.0rc1.post1 Requires-Python >=3.6,<3.10; 0.53.0rc2 Requires-Python >=3.6,<3.10; 0.53.0rc3 Requires-Python >=3.6,<3.10; 0.53.1 Requires-Python >=3.6,<3.10; 0.54.0 Requires-Python >=3.7,<3.10; 0.54.0rc2 Requires-Python >=3.7,<3.10; 0.54.0rc3 Requires-Python >=3.7,<3.10; 0.54.1 Requires-Python >=3.7,<3.10; 0.55.0 Requires-Python >=3.7,<3.11; 0.55.0rc1 Requires-Python >=3.7,<3.11; 0.55.1 Requires-Python >=3.7,<3.11; 0.55.2 Requires-Python >=3.7,<3.11
> > ERROR: Could not find a version that satisfies the requirement dolomite-engine (from instructlab-training) (from versions: none)
> > ERROR: No matching distribution found for dolomite-engine
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Found this when pip installing, I think this will fail the linter as well
> 
> @RobotSail @Maxusmusti @JamesKunstle any idea why this is happening? I'm hitting the same issue



> > ```
> > ERROR: Ignored the following versions that require a different python version: 0.52.0 Requires-Python >=3.6,<3.9; 0.52.0rc3 Requires-Python >=3.6,<3.9; 0.53.0 Requires-Python >=3.6,<3.10; 0.53.0rc1.post1 Requires-Python >=3.6,<3.10; 0.53.0rc2 Requires-Python >=3.6,<3.10; 0.53.0rc3 Requires-Python >=3.6,<3.10; 0.53.1 Requires-Python >=3.6,<3.10; 0.54.0 Requires-Python >=3.7,<3.10; 0.54.0rc2 Requires-Python >=3.7,<3.10; 0.54.0rc3 Requires-Python >=3.7,<3.10; 0.54.1 Requires-Python >=3.7,<3.10; 0.55.0 Requires-Python >=3.7,<3.11; 0.55.0rc1 Requires-Python >=3.7,<3.11; 0.55.1 Requires-Python >=3.7,<3.11; 0.55.2 Requires-Python >=3.7,<3.11
> > ERROR: Could not find a version that satisfies the requirement dolomite-engine (from instructlab-training) (from versions: none)
> > ERROR: No matching distribution found for dolomite-engine
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Found this when pip installing, I think this will fail the linter as well
> 
> @RobotSail @Maxusmusti @JamesKunstle any idea why this is happening? I'm hitting the same issue

`dolomite_engine` does not appear to be a valid package that exists on pypi

**Comment by cdoern** [comment]
also when bringing this into the cli:

```
ERROR: Cannot install instructlab, instructlab-training==0.1.dev22 and instructlab==0.16.1.dev140 because these package versions have conflicting dependencies.

The conflict is caused by:
    instructlab 0.16.1.dev140 depends on transformers<=4.38.2 and >=4.30.0
    peft 0.9.0 depends on transformers
    trl 0.7.11 depends on transformers>=4.31.0
    instructlab-training 0.1.dev22 depends on transformers==4.41.2

```

have these versions been vetted as ok to bump in the cli?

**Comment by nathan-weinberg** [comment]
> also when bringing this into the cli:
> 
> ```
> ERROR: Cannot install instructlab, instructlab-training==0.1.dev22 and instructlab==0.16.1.dev140 because these package versions have conflicting dependencies.
> 
> The conflict is caused by:
>     instructlab 0.16.1.dev140 depends on transformers<=4.38.2 and >=4.30.0
>     peft 0.9.0 depends on transformers
>     trl 0.7.11 depends on transformers>=4.31.0
>     instructlab-training 0.1.dev22 depends on transformers==4.41.2
> ```
> 
> have these versions been vetted as ok to bump in the cli?

Can this be a separate issue/convo than the linting introduction?

**Review by cdoern — APPROVED** [review]

File: requirements.txt [hunk 3]
```diff
@@ -9,7 +9,7 @@ datasets
 numba
 numpy
 rich
-dolomite_engine
+git+https://github.com/ibm-granite/dolomite-engine.git
```
Comment (RobotSail) [inline requirements.txt:12]: Nice, I didn't know we could do this!

**Review by RobotSail — APPROVED** [review]
This looks good, LGTM

**Review by russellb — APPROVED** [review]
Thanks, @nathan-weinberg !

**Comment by russellb** [comment]
Well, I thought we were good.  Most stuff works, but `tox -e mypy` is busted on the requirements.txt change made for dolomite_engine.

```
❯ tox -e mypy
.pkg: _optional_hooks> python /Users/rbryant/src/instructlab/venv/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_sdist> python /Users/rbryant/src/instructlab/venv/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta
mypy: packaging backend failed (code=1), with InvalidRequirement: Expected end or semicolon (after name and no valid version specifier)
    git+https://github.com/ibm-granite/dolomite-engine.git
```

**Comment by russellb** [comment]
> Well, I thought we were good. Most stuff works, but `tox -e mypy` is busted on the requirements.txt change made for dolomite_engine.
> 
> ```
> ❯ tox -e mypy
> .pkg: _optional_hooks> python /Users/rbryant/src/instructlab/venv/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta
> .pkg: get_requires_for_build_sdist> python /Users/rbryant/src/instructlab/venv/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta
> mypy: packaging backend failed (code=1), with InvalidRequirement: Expected end or semicolon (after name and no valid version specifier)
>     git+https://github.com/ibm-granite/dolomite-engine.git
> ```

I got this working and pushed it to the branch

**Comment by russellb** [comment]
I'm going to merge and we'll keep working on this if anything doesn't pass right away

**Pull request merged**